.github/workflows/agent-build-refactored.yml

```diff
@@ -0,0 +1,166 @@
+name: Agent Build Refactored
+
+on:
+  push:
+  pull_request:
+    branches:
+      - master
+  schedule:
+    - cron: '0 4 * * *'
+
+permissions:
+  contents: read
+
+env:
+  DOCKER_BUILDKIT: 1
+  # Set this variable to tell the agent build code that it runs in CI/CD and it needs to use caching.
+  AGENT_BUILD_IN_CICD: "1"
+
+# This job skips another workflows with identical content and also generates strategy matrices for all further jobs.
+# Since we do a "limited" run on non-master custom branch, it generates matrices with different size according to
+# that information.
+jobs:
+  pre-job:
+    runs-on: ubuntu-20.04
+    outputs:
+      should_skip: ${{ steps.skip_check.outputs.should_skip }}
+      # The following outputs will be used by the other jobs below as their strategy matrices.
+      pre_build_steps_matrix_json: ${{ steps.init-matrices.outputs.pre_build_steps_matrix_json }}
+      agent_image_build_matrix_json: ${{ steps.init-matrices.outputs.agent_image_build_matrix_json }}
+
+    steps:
+      - id: skip_check
+        uses: fkirc/skip-duplicate-actions@9d116fa7e55f295019cfab7e3ab72b478bcf7fdd # v4.0.0
+        with:
+          cancel_others: 'true'
+          concurrent_skipping: 'same_content'
+          do_not_skip: '["workflow_dispatch", "schedule", "pull_request"]'
+          github_token: ${{ github.token }}
+
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Install python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.8.13"
+
+      # In this action we define all our job matrices. It filters out jobs that must not be executed if
+      # workflow run is not from the 'master' branch or from PR to the 'master' branch.
+      # Jobs that has to be in a non-master workflow run has to have a '"basic": true' field.
+      - name: Init job matrices
+        uses: ./.github/actions/init-job-matrices
+        id: init-matrices
+        with:
+
+          # In the 'non-master' runs we only build debian-based images.
+          images-to-build: |
+            [
+              { "name": "docker-json-debian", "basic": true },
```

Comment:
I know there is explanation in the comment above which explains what "basic" means, but I was still confused a bit before checking the comment.

I'm just thinking out loud if we could come up with some more explanatory attribute name - but maybe that's not possible since I can't think of anything better either :)

I guess another way would perhaps be to change this behavior and add ``master_run_only`` (or similar) attribute and set it to true for alpine jobs and false for debian ones.